backend/src/api/realms.rs

```diff
@@ -0,0 +1,95 @@
+use std::collections::HashMap;
+use futures::{stream::TryStreamExt, TryStream};
+
+use juniper::graphql_object;
+
+use super::Context;
+use deadpool_postgres::Pool;
+
+
+pub(super) type Id = i32;
+
+pub(super) struct Realm {
+    id: Id,
+    name: String,
+    parent_id: Id,
+}
+
+#[graphql_object(Context = Context)]
+impl Realm {
+    fn id(&self) -> &Id {
+        &self.id
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn parent_id(&self) -> &Id {
+        &self.parent_id
+    }
+
+    fn parent(&self, context: &Context) -> &Realm {
+        &context.realm_tree.realms[&self.parent_id]
+    }
+
+    fn children(&self, context: &Context) -> Vec<&Realm> {
+        if let Some(children) = context.realm_tree.children.get(&self.id) {
+            children.iter().map(
+                |child_id| &context.realm_tree.realms[&child_id],
+            ).collect()
+        } else {
+            vec![]
+        }
+    }
+}
+
+pub(super) struct Tree {
+    pub(super) realms: HashMap<Id, Realm>,
+    children: HashMap<Id, Vec<Id>>,
+}
+
+impl Tree {
+    pub(super) async fn raw_from_db(
+        db: &Pool,
+    ) -> anyhow::Result<impl TryStream<Ok = Realm, Error = impl std::error::Error>> {
+        let row_stream = db.get().await?
+            .query_raw(
+                "select id, name, parent from realms",
+                std::iter::empty(),
+            ).await?;
+        Ok(row_stream.map_ok(|row| Realm {
+            id: row.get(0),
+            name: row.get(1),
+            parent_id: row.get(2),
+        }))
+    }
+
+    pub(super) async fn from_db(db: &Pool) -> anyhow::Result<Self> {
+        // We store the nodes of the realm tree in a hash map
+        // accessible by the database ID
+        let realms = Self::raw_from_db(db).await?
+            .map_ok(|realm| (
+                realm.id,
+                realm,
+            ))
```

Comment:
No hard opinion; I just thought this would highlight the structure of what is returned here more explicitly: We have a collection (highlighted by the `map` in one line) of tuples with suchandsuch members (all on their own lines as well. :thinking:

Reply:
It's just really short in this case, so I would say splitting into multiple lines is not necessary.